vault/namespaces_store.go

```diff
@@ -734,54 +755,187 @@ func clearNamespaceResources(ctx context.Context, ns *NamespaceStore, namespaceT
+	err := logical.WithTransaction(ctx, ns.storage, func(s logical.Storage) error {
+		item, err := s.Get(ctx, lockPath)
+		if err != nil {
+			return err
+		}
+
+		if item == nil {
+			return errors.New("couldn't find namespace lock in storage")
+		}
+
+		return item.DecodeJSON(&nsLock)
+	})
+	if err != nil {
+		return fmt.Errorf("error retrieving namespace lock: %w", err)
+	}
+
+	// verify lock or omit when forced unlock
+	if !forceUnlock && nsLock.Key != unlockKey {
+		return errors.New("incorrect unlock key")
+	}
+
+	// delete lock from storage
+	err = logical.WithTransaction(ctx, ns.storage, func(s logical.Storage) error {
+		return s.Delete(ctx, lockPath)
+	})
 	if err != nil {
-		return intoCtx, nil, fmt.Errorf("could not fetch namespace by path: %w", err)
+		return fmt.Errorf("error deleting namespace lock: %w", err)
 	}
 
-	if entry == nil {
-		return intoCtx, nil, fmt.Errorf("requested namespace was not found")
+	ns.namespacesByAccessor[namespace.ID].Locked = false
+	ns.namespacesByUUID[namespace.UUID].Locked = false
+	namespace.Locked = false
+	err = ns.namespacesByPath.Insert(namespace)
+	if err != nil {
+		return fmt.Errorf("failed to modify namespace tree: %w", err)
 	}
 
-	intoCtx = namespace.ContextWithNamespace(intoCtx, entry)
-	return intoCtx, entry, nil
+	return nil
 }
 
-// ResolveNamespaceFromRequest merges the given base context with the
-// namespace from httpCtx, combining it with any namespaces within the
-// request path itself. We remove the prefix from the path, if given,
-// because logic elsewhere in vault/ combines the namespace with the
-// path again.
-func (ns *NamespaceStore) ResolveNamespaceFromRequest(baseCtx context.Context, httpCtx context.Context, reqPath string) (context.Context, *namespace.Namespace, string, error) {
-	ns.lock.RLock()
-	defer ns.lock.RUnlock()
+// LockNamespace attempts to lock the namespace with provided uuid.
+func (ns *NamespaceStore) LockNamespace(ctx context.Context, uuid string) (string, error) {
+	defer metrics.MeasureSince([]string{"namespace", "lock_namespace"}, time.Now())
+
+	if err := ns.checkInvalidation(ctx); err != nil {
+		return "", err
+	}
 
-	// We stack the namespace context ahead of any namespace in path.
-	newCtx, parentNs, err := ns.copyNamespaceFromCtx(baseCtx, httpCtx)
+	namespaceToLock, err := ns.GetNamespace(ctx, uuid)
 	if err != nil {
-		return newCtx, parentNs, reqPath, err
+		return "", err
+	}
+
+	if namespaceToLock == nil {
+		return "", nil
 	}
 
-	// prepend namespace path from request context
-	reqPath = parentNs.Path + reqPath
-	// find namespace that matches the longest prefix of reqPath
-	nsPath, parentNs, _ := ns.namespacesByPath.LongestPrefix(reqPath)
-	// trim matched prefix from reqPath
-	reqPath = reqPath[len(nsPath):]
+	lockedNamespace := ns.GetLockingNamespace(namespaceToLock)
+	if lockedNamespace != nil {
+		return "", fmt.Errorf("cannot lock namespace %q as %q is already locked", namespaceToLock.Path, lockedNamespace.Path)
+	}
```

Comment:
Maybe to make it even clearer to the user

```suggestion
	lockedNamespace := ns.GetLockingNamespace(namespaceToLock)
	if lockedNamespace.ID == namespaceToLock.ID {
		return "", fmt.Errorf("cannot lock namespace %q: already locked", namespaceToLock.Path)
	} else if lockedNamespace != nil {
		return "", fmt.Errorf("cannot lock namespace %q: parent namespace %q is already locked", namespaceToLock.Path, lockedNamespace.Path)
	}
```